### ISSUE TITLE
perf+compat: a37641b regresses JAX 0.8.x by +0.38s, breaks JAX ≤0.7.x compat

### ISSUE THREAD

**Issue by tinker495** [tinker495/Xtructure#5]
# [perf/compat] `a37641b` (single fingerprint + insert_triton.py) — JAX 0.8.x 위 +0.38s 회귀 + JAX ≤0.7.x 호환성 손실

## 요약 (TL;DR)

`a37641b81ca6e28f232ebb8a4171306041fb019c` (2026-01-25, "single fingerprint hashtable lookup + insert_triton.py 도입") 는 두 가지 문제를 동시에 일으킵니다:

1. **JAX 0.7.x 호환성 손실** — 신규 `xtructure/hashtable/insert_triton.py:107` 에서 사용하는 `jax.experimental.pallas.triton.atomic_or` 가 jax ≤ 0.7.x 에는 존재하지 않습니다. JAxtar 같은 다운스트림 사용자가 jax 0.7.x 에서 import 하자마자 `AttributeError` 로 크래시.
2. **JAX 0.8.x 위에서 +0.38s wall-clock 회귀** — JAxtar A* DeepCubeA 벤치마크에서 같은 명령 + 같은 docs-era JAxtar/PuXle 위에서 직전 SHA `2b39104` (1.370s) → `a37641b` (1.754s).

JAX 0.9.x 위에서는 (3.4 측정) 이 회귀가 가려져서 `eee7a6b` 와 `45723c1` 사이 차이가 noise (~0.018s) 수준입니다. 이는 JAX 0.9.x 가 두 코드 경로 모두를 비싸게 컴파일하기 때문이고, JAX 0.8.x 에서는 신규 경로 (a37641b+) 가 구경로보다 ~+0.3s 더 비쌉니다.

이 두 문제는 함께 다음 영향을 만듭니다:
- 다운스트림은 JAX 0.7.x 위에서 Xtructure HEAD 를 쓸 수 없음 (호환성)
- 다운스트림은 JAX 0.8.x 위에서 같은 워크로드가 +0.38s 느려진 것을 그대로 떠안음 (성능)
- 회귀를 우회하려면 Xtructure 를 `2b39104` 이전 SHA 로 핀해야 하는데, 이는 모듈 분리 이전 코드라 새 기능 사용처에 영향

### JAX 0.8.1 × Xtructure SHA sweep (JAxtar=`4fb2a800`, PuXle=`8f50034` 고정)

| xt SHA | 날짜 | mean wall-clock (s) | Δ vs prev |
| :--- | :--- | ---: | ---: |
| `eee7a6b` | 2025-12-19 | 1.484 | (anchor) |
| `6fbdf2e` | 2026-01-11 | 1.427 | -0.057 |
| `080b291` | 2026-01-11 | 1.396 | -0.031 |
| `b1e9a4f` | 2026-01-12 | 1.457 | +0.061 |
| `13aeebb` | 2026-01-14 | 1.428 | -0.029 |
| `2b39104` | 2026-01-24 | 1.370 | -0.058 |
| **`a37641b`** | **2026-01-25** | **1.754** | **+0.384** ⚡ |
| `e8d332a` | 2026-01-25 | 1.868 | +0.114 |
| `5508b91` | 2026-02-22 | 1.789 | -0.079 |
| `1a17fd1` | 2026-03-02 | 1.744 | -0.045 |
| `45723c1` (HEAD) | 2026-03-03 | 1.775 | +0.031 |

**클러스터 분리:** early 평균 1.427s (n=12 trials), late 평균 1.786s (n=10 trials). Δ=0.359s. 클러스터 내 표준편차 ≤ 0.06s, 클러스터 간 분리 5σ 이상.

### JAX 0.7.0 × pre-`atomic_or` Xtructure SHA (JAxtar=`4fb2a800`, PuXle=`8f50034` 고정)

| xt SHA | mean (s) |
| :--- | ---: |
| `eee7a6b` | 1.413 |
| `6fbdf2e` | 1.371 |
| `080b291` | 1.335 |
| `b1e9a4f` | 1.298 ⭐ |
| `13aeebb` | 1.345 |
| `2b39104` | 1.364 |

JAX 0.7.0 + `b1e9a4f` 조합이 전체 측정 중 가장 빠름 (`1.298s`). docs row 보고치 `1.438s` 보다도 빠르며, JAxtar HEAD + xt HEAD + jax 0.9.x 의 `~2.20s` 와 비교하면 약 41% 차이.

**`a37641b` 이후 SHA 는 jax 0.7.x 에서 비측정** — `pl_triton.atomic_or` 가 jax 0.7.x 에 미존재이기 때문.

### JAX 0.9.0.1 × Xtructure SHA sample (JAxtar=`4fb2a800`, PuXle=`8f50034` 고정)

| xt SHA | mean (s) |
| :--- | ---: |
| `eee7a6b` | 2.140 |
| `a37641b` | 2.265 |
| `5508b91` | 2.141 |
| `45723c1` | 2.158 |

JAX 0.9.x 위에서는 `a37641b` 의 +0.38s 효과가 +0.12s 로 줄어들고, 다른 SHA 들은 noise 안. **JAX 0.9.x 가 두 코드 경로를 같은 (비싼) 길로 컴파일한다는 강력한 정황 증거.**

## 재현

```bash
# Setup (한 번만)
git -C /path/to/Xtructure worktree add /tmp/xt-test eee7a6b
conda create -n test-jax081 --clone py312
conda activate test-jax081
pip install jax[cuda12]==0.8.1
pip install -e /tmp/xt-test
pip install -e /path/to/PuXle  # 8f50034 추천

# 측정 1: pre-a37641b
git -C /tmp/xt-test checkout 2b39104
cd /path/to/JAxtar  # 4fb2a800 worktree
XLA_PYTHON_CLIENT_PREALLOCATE=false XTRUCTURE_BGPQ_MERGE_BACKEND=sort \
  python main.py benchmark astar --benchmark rubikscube-deepcubea \
  --sample-ids 0,1,2,3,4,5,6,7,8,9 --batch-size 10000 \
  --max-node-size 20000000 --cost-weight 0.6 --pop_ratio inf \
  --use-early-stopping False
# Expected: Avg. Search Time (Solved) ≈ 1.37s

# 측정 2: post-a37641b
git -C /tmp/xt-test checkout a37641b
cd /path/to/JAxtar
XLA_PYTHON_CLIENT_PREALLOCATE=false XTRUCTURE_BGPQ_MERGE_BACKEND=sort \
  python main.py benchmark astar --benchmark rubikscube-deepcubea \
  --sample-ids 0,1,2,3,4,5,6,7,8,9 --batch-size 10000 \
  --max-node-size 20000000 --cost-weight 0.6 --pop_ratio inf \
  --use-early-stopping False
# Expected: Avg. Search Time (Solved) ≈ 1.75s
```

## 환경

- Python 3.12.12, NumPy 2.4.2, Flax 0.12.2, Optax 0.2.8.dev0
- GPU: NVIDIA RTX 4080 SUPER, driver 591.86
- CUDA: 12.x (jax-cuda12-pjrt / jax-cuda12-plugin)
- `XLA_PYTHON_CLIENT_PREALLOCATE=false`
- `XTRUCTURE_BGPQ_MERGE_BACKEND=sort` (a37641b 이전 SHA 의 docs-era Pallas BLOCK_SIZE=64 가 현재 GPU 의 SMEM 한계 위반하는 것 우회용; 클러스터 분리에는 영향 없음)

### Priority 1 — JAX 0.7.x 호환성 회복

`pl_triton.atomic_or` 가 미존재할 경우 자동 fallback 하는 import 가드:

```python
# xtructure/hashtable/insert_triton.py 상단
from jax.experimental.pallas import triton as pl_triton

if not hasattr(pl_triton, "atomic_or"):
    raise ImportError(
        "jax.experimental.pallas.triton lacks atomic_or. "
        "Use XTRUCTURE_HASHTABLE_INSERT_TRITON=0 to disable triton path, "
        "or upgrade jax to >=0.8.0."
    )
```

또는 더 견고하게: `parallel_insert` 가 `_HAS_TRITON_ATOMIC_OR` 플래그를 체크하고, false 면 기존 (non-triton) 경로로 fallback.

### Priority 2 — `insert_triton.py` 경로 opt-in 으로 변경

env var `XTRUCTURE_HASHTABLE_INSERT_TRITON=1` (default `0`) 로 새 경로 opt-in.

근거: JAX 0.8.x 위에서 +0.38s 회귀가 dominant. JAX 0.9.x 위에서는 어차피 가려지므로 새 경로의 이득이 측정되지 않음. 새 경로가 어떤 워크로드/JAX 버전 위에서 명확한 성능 이득을 만드는지 별도 벤치 후 default 결정.

### Priority 3 — 회귀 보고 (메타)

이 회귀는 같은 commit 의 세 변경 (`single fingerprint refactor` + `insert_triton.py 도입` + `lru_cache on dataclass property`) 중 어느 것이 dominant 인지 추가 분리 측정 필요. 후속 issue / PR 후보:

- (a) `single fingerprint refactor` 만 retain, `insert_triton.py` 만 revert 한 commit 측정
- (b) 위와 반대
- (c) `lru_cache` 만 revert

## 관련 산출물

- 통합 보고서: 다운스트림 JAxtar 의 `JAX_XTRUCTURE_REGRESSION_REPORT.md`
- Raw 데이터: `<JAxtar repo>/.omc/autoresearch/jax-version-axis-validation/runs/20260504-005613/extension_xt_sweep/`

## 라벨 제안

`bug`, `performance`, `compatibility`, `jax-version`
